### PR TITLE
Token Transfer support contract address

### PR DESCRIPTION
**Pull request by reedrosenbluth**
Updates the TokenTransfer transaction payload to use a Principal instead of an address for the recipient. This is enables sending STX to either a standard principal or a contract principal. 

Issue: #24
Related: https://github.com/blockstack/stacks-blockchain/issues/1317

## Type of Change
- [x] New feature
- [ ] Bug fix
- [ ] API reference/documentation update
- [ ] Other

## Checklist
- [x] Code is commented where needed
- [x] Unit test coverage for new or modified code paths
- [x] `npm run test` passes
- [ ] Changelog is updated
- [x] Tag 1 of @yknl, @zone117x, @reedrosenbluth for review

### CODE REVIEW

File: src/payload.ts [hunk 1]
```diff
@@ -31,27 +38,24 @@ export type Payload =
...
   readonly payloadType: PayloadType.TokenTransfer;
-  readonly recipientAddress: Address;
+  readonly recipient: PrincipalCV;
   readonly amount: BigNum;
   readonly memo: MemoString;
 }
 
 export function createTokenTransferPayload(
-  recipientAddress: string | Address,
+  recipient: PrincipalCV,
```
Comment (zone117x) [inline src/payload.ts:47]: I think for ease of use this should still take a string `recipient: PrincipalCV | string`

File: src/postcondition.ts [hunk 2]
```diff
@@ -28,13 +27,13 @@ export type PostCondition = STXPostCondition | FungiblePostCondition | NonFungib
...
   readonly conditionType: PostConditionType.STX;
-  readonly principal: Principal;
+  readonly principal: PostConditionPrincipal;
   readonly conditionCode: FungibleConditionCode;
   readonly amount: BigNum;
 }
 
 export function createSTXPostCondition(
-  principal: Principal,
+  principal: PostConditionPrincipal,
```
Comment (zone117x) [inline src/postcondition.ts:36]: Same here (accepting a `string` type as well)

File: src/postcondition.ts [hunk 3]
```diff
@@ -75,14 +74,14 @@ export function createFungiblePostCondition(
...
-  readonly principal: Principal;
+  readonly principal: PostConditionPrincipal;
   readonly conditionCode: NonFungibleConditionCode;
   readonly assetInfo: AssetInfo;
   readonly assetName: LengthPrefixedString;
 }
 
 export function createNonFungiblePostCondition(
-  principal: Principal,
+  principal: PostConditionPrincipal,
```
Comment (zone117x) [inline src/postcondition.ts:84]: Same here (accepting a `string` type as well)

File: tests/src/payload-tests.ts [hunk 4]
```diff
@@ -11,25 +11,24 @@ import {
...
+import { trueCV, falseCV, standardPrincipalCV } from '../../src/clarity';
 
 import * as BigNum from 'bn.js';
 
 import { COINBASE_BUFFER_LENGTH_BYTES, StacksMessageType } from '../../src/constants';
-import { addressToString } from '../../src';
 
 test('STX token transfer payload serialization and deserialization', () => {
-  const recipientAddress = 'SP3FGQ8Z7JY9BWYZ5WM53E0M9NK7WHJF0691NZ159';
+  const recipient = standardPrincipalCV('SP3FGQ8Z7JY9BWYZ5WM53E0M9NK7WHJF0691NZ159');
```
Comment (zone117x) [inline tests/src/payload-tests.ts:21]: Should we have a test for `STX token transfer to contract principal`?

**Comment by reedrosenbluth** [comment]
@zone117x great idea, looks good!

**Pull request merged**